internal/server/server.go

```diff
@@ -1,2 +1,179 @@
+// This package containts server implementation to receive events in format described in
+// thermomatic/internal/client/reading.go.
 package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/spin-org/thermomatic/internal/client"
+	"github.com/spin-org/thermomatic/internal/imei"
+)
+
+const (
+	loginTimeout = 1 * time.Second
+	readTimeout = 2 * time.Second
+)
+
+var errReadTimeout = errors.New("read timeout")
+
+type Server struct {
+	create chan net.Conn
+	exit   chan net.Conn
+	clients map[net.Conn]struct{}
+
+	stdout io.Writer
+}
+
+func New(stdout io.Writer) *Server {
+	return &Server{
+		create: make(chan net.Conn),
+		exit:   make(chan net.Conn),
+
+		clients: make(map[net.Conn]struct{}),
+		stdout: stdout,
+	}
+}
+
+func (s *Server) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	lc := net.ListenConfig{}
+	l, err := lc.Listen(ctx, "tcp4", ":1337")
+	if err != nil {
+		return err
+	}
+
+	go s.accept(l)
+
+outer:
+	for {
+		select {
+		case conn := <- s.create:
+			log("%v %v new connection established\n", conn, conn.RemoteAddr())
+			s.clients[conn] = struct{}{}
+			go s.handleClient(conn)
+		case conn := <-s.exit:
+			log("%v %v connection is closing\n", conn, conn.RemoteAddr())
+			delete(s.clients, conn)
+			_ = conn.Close()
+		case <- ctx.Done():
+			log("graceful shutdown\n")
+			_ = l.Close()
+			for c, _ := range s.clients {
+				_ = c.Close()
+			}
+			break outer
+		}
+	}
+	s.shutdown()
+	return nil
+}
+
+func (s *Server) shutdown() {
+	if len(s.clients) == 0 {
+		return
+	}
+
+	for {
+		select {
+			case conn := <- s.create:
+				_ = conn.Close()
+			case conn := <- s.exit:
+				delete(s.clients, conn)
+				if len(s.clients) == 0 {
+					return
+				}
+		}
+	}
+}
+
+func (s *Server) accept(l net.Listener) {
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			return
+		}
+		if conn != nil {
+			s.create <- conn
+		}
+	}
+}
+
+func (s *Server) handleClient(conn net.Conn) {
+	code, err := s.login(conn)
+	if err != nil {
+		log("%v %v login error: %v\n", conn, conn.RemoteAddr(), err)
+		s.exit <- conn
+		return
+	}
+
+	err = s.read(code, conn)
+	if err != nil {
+		log("%v %v read data error: %v\n", conn, conn.RemoteAddr(), err)
+	}
+	s.exit <- conn
+}
+
+func (s *Server) login(conn net.Conn) (code uint64, err error) {
+	err = conn.SetReadDeadline(time.Now().Add(loginTimeout))
+	if err != nil {
+		return
+	}
+
+	var i [15]byte
+	_, err = io.ReadFull(conn, i[:])
+	if err != nil {
+		if _, ok := err.(*net.OpError); ok {
+			err = errReadTimeout
+		}
+		return
+	}
+	return imei.Decode(i[:])
+}
+
+func (s *Server) read(code uint64, conn net.Conn) (err error) {
+	var data [40]byte
+	r := &client.Reading{}
+	for {
+		if err = conn.SetReadDeadline(time.Now().Add(readTimeout)); err != nil {
+			return
+		}
+
+		_, err = io.ReadFull(conn, data[:])
+		if err != nil {
+			if _, ok := err.(*net.OpError); ok {
```

Comment:
is `*net.OpError` always a timeout?

Reply:
Of course not. net.OpError can wrap syscall.Errno as well. I could not find an easy way to determine if it was timeout or not. On unix like system of course there is syscall errno ETIMEDOUT  but of course that doesn't work on mac system.

Reply:
https://golang.org/pkg/net/#OpError.Timeout

Reply:
Ops, *net.OpError has Timeout() method